File: iOS/Schema/Sources/Widgets/Image/Image.swift
/*
 * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import Foundation

public struct Image: RawWidget, AutoDecodable {

    // MARK: - Public Properties
    public let path: Expression<ImagePath>
    public let mode: ImageContentMode?
    public var widgetProperties: WidgetProperties
    
    public init(
        _ path: Expression<ImagePath>,
        mode: ImageContentMode? = nil,
        widgetProperties: WidgetProperties = WidgetProperties()
    ) {
        self.path = path
        self.mode = mode
        self.widgetProperties = widgetProperties
    }
    
    public init(
        _ path: ImagePath,
        mode: ImageContentMode? = nil,
        widgetProperties: WidgetProperties = WidgetProperties()
    ) {
        self.path = .value(path)
        self.mode = mode
        self.widgetProperties = widgetProperties
    }
    
    public enum ImagePath: Decodable {
        case remote(Remote)
        case local(StringOrExpression)

        enum CodingKeys: String, CodingKey {
            case type = "_beagleImagePath_"
            case mobileId
        }
        
        public init(from decoder: Decoder) throws {
            let container = try decoder.container(keyedBy: CodingKeys.self)

            let type = try container.decode(String.self, forKey: .type)
            switch type {
            case "local":
                let mobileId = try container.decode(String.self, forKey: .mobileId)
                self = .local(mobileId)
            case "remote":
                self = .remote(try Remote(from: decoder))
            default:
                throw DecodingError.dataCorruptedError(forKey: .type, in: container, debugDescription: "Invalid image type")
            }
        }
    }
}

public extension Image {
    struct Remote: Decodable {
        public let url: StringOrExpression
        public let placeholder: String?

        public init(url: StringOrExpression, placeholder: String? = nil) {
            self.url = url
            self.placeholder = placeholder
        }

        enum CodingKeys: String, CodingKey {
            case url
            case placeholder
        }
        
        enum LocalImageCodingKey: String, CodingKey {
            case mobileId
        }

        public init(from decoder: Decoder) throws {
            let container = try decoder.container(keyedBy: CodingKeys.self)
            let nestedContainer = try? container.nestedContainer(keyedBy: LocalImageCodingKey.self, forKey: .placeholder)
            url = try container.decode(String.self, forKey: .url)
            placeholder = try nestedContainer?.decodeIfPresent(String.self, forKey: .mobileId)
        }
    }
}

File: backend/backend-widgets/src/main/kotlin/br/com/zup/beagle/widget/ui/Image.kt
/*
 * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package br.com.zup.beagle.widget.ui

import br.com.zup.beagle.widget.Widget
import br.com.zup.beagle.widget.context.Bind
import br.com.zup.beagle.widget.context.valueOf
import br.com.zup.beagle.widget.core.ImageContentMode

/**
 * Define an image view using the server driven information received through Beagle.
 *
 * @param path defines where the source of the image is
 * @param mode defines how the declared image will fit the view.
 */
data class Image(val path: Bind<ImagePath>, val mode: ImageContentMode? = null) : Widget() {
    constructor(path: ImagePath, mode: ImageContentMode? = null) : this(valueOf(path), mode)
}


/**
 * Define the source of image data to populate the image view.
 *
 * @param url the URL from which the image can be obtained
 * @param placeholder the image that will used as placeholder when set a remote image.
 * */
sealed class ImagePath(val url: String?, val placeholder: Local? = null) {
    /**
     * Define an image whose data is local to the client app.
     *
     * @param webUrl reference the path to an image in your web app files.
     * @param mobileId reference an image natively in your mobile app local styles file.
     * */
    class Local internal constructor(webUrl: String?, val mobileId: String?) : ImagePath(webUrl) {
        companion object {
            fun both(webUrl: String, mobileId: String) = Local(webUrl, mobileId)
            fun justMobile(mobileId: String) = Local(null, mobileId)
            fun justWeb(webUrl: String) = Local(webUrl, null)
        }
    }

    /**
     * Define an image whose data needs to be downloaded from a source external to the client app.
     *
     * @param remoteUrl reference the path where the image should be fetched from.
     * @param placeholder reference an image natively in your mobile app local styles file to be used as placeholder.
     * */
    class Remote(remoteUrl: String, placeholder: Local? = null) : ImagePath(remoteUrl, placeholder)
}